streaming/stream_transfer_task.cc

```diff
@@ -199,18 +199,27 @@ future<> send_mutation_fragments(lw_shared_ptr<send_info> si) {
         }();
 
         auto sink_op = [sink, si, got_error_from_peer] () mutable -> future<> {
-            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment>& sink) {
+            return do_with(std::move(sink), [si, got_error_from_peer] (rpc::sink<frozen_mutation_fragment, stream_mutation_fragments_cmd>& sink) {
                 return repeat([&sink, si, got_error_from_peer] () mutable {
                     return si->reader(db::no_timeout).then([&sink, si, s = si->reader.schema(), got_error_from_peer] (mutation_fragment_opt mf) mutable {
                         if (mf && !(*got_error_from_peer)) {
                             frozen_mutation_fragment fmf = freeze(*s, *mf);
                             auto size = fmf.representation().size();
                             streaming::get_local_stream_manager().update_progress(si->plan_id, si->id.addr, streaming::progress_info::direction::OUT, size);
-                            return sink(fmf).then([] { return stop_iteration::no; });
+                            return sink(fmf, stream_mutation_fragments_cmd::mutation_fragment_data).then([] { return stop_iteration::no; });
                         } else {
                             return make_ready_future<stop_iteration>(stop_iteration::yes);
                         }
                     });
+                }).then([&sink] () mutable {
+                    return sink(frozen_mutation_fragment(bytes_ostream()), stream_mutation_fragments_cmd::end_of_stream);
+                }).then([&sink] () mutable {
+                    return sink.flush();
+                }).handle_exception([&sink] (std::exception_ptr ep) mutable {
+                    // Notify the receiver the sender has failed
+                    return sink(frozen_mutation_fragment(bytes_ostream()), stream_mutation_fragments_cmd::error).then([ep = std::move(ep)] () mutable {
```

Comment:
Sending `error` doesn't seem very useful. If we already sent end_of_stream, we can as well let the stream succeed. If we didn't send end_of_stream, the receiver will deduce there was an error.

Reply:
Why deducing instead explicitly telling the peer there is an error? 

If we sent the error explicitly the peer can lean the error faster, for example:

Sender:
1. error happened
2. send error
3. send close

Skipping 2), the receiver only knows there is error when the stream is closed.  There can be long delay between 1) and 3), e.g., sudden lost of the node.

end_of_stream is sent lastly only if there is nothing to send. If the receiver receives end_of_stream, it means the stream is end successfully. It will not care about  whatever was sent latter.

Reply:
We already deduce the error when the stream is closed. Sending `error` is only extra complexity and extra code.

sending close is the same as sending error. Sending close is not delayed more than sending error is.

Reply:
I do not understand what was wrong with the version that had an error only. Having explicit error notification allows passing assitional info about an error an (error code, or an error message). The patch does not do it, but it can be extended if needed.

Reply:
Passed additional info would be useful indeed. But in this patch the error message is just an enum value, so it cannot do that nor can be extended later. As such, it's a dead end and unnecessary.

Reply:
It's not that of a big deal though. It may be useful for distinguishing sender failure from other possible (?) reasons of premature end of stream.